doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,196 @@
+* `WriteCapability(id, frontier)` prevents the `upper` frontier associated with `id` from advancing beyond `frontier`.
+
+A capability presents the `id` and `frontier` values, but is identified by a system-assigned unique identifier.
+Any user presenting this identifier has the associated capability to read from or write to the associated pTVC.
+Actions which read or write will validate the capability (e.g. in case a lease has expired) and may return errors if the capability is not valid as stated.
+
+Capabilities are durable, and their state is tracked by the system who associates a unique identifier with each.
+Capabilities can be cloned and transfered to others, providing the same rights (and responsibilities).
+Capabilities can be downgraded to frontiers that are greater or equal to the held frontier, which may then allow the `since` or `upper` frontier to advance.
+Capabilities can be dropped, which releases the constraint on the frontier, and the whichever ability (read or write) the capability provided.
+These actions are mediated by the system, and take effect only when the system confirms the action.
+Users can defer and batch these actions, modulo lease timeouts, imposing only on the freshness and efficiency of the system.
+
+Capabilities may expire, as a lease, which may result in errors from attempts to exercise the capability.
+
+It is very common for users to immediately drop capabilities if they do not require the ability.
+The write capability for most TVCs is dropped immediately (by the system) unless it is meant to be written to by users (e.g. a `TABLE`).
+The read capability may be dropped immediately if the user does not presently need to read the data.
+Capabilities may be recovered by the best-effort command `AcquireCapabilities(id)` to be discussed.
+
+## Storage
+
+The Storage layer presents a narrow API about which it makes guarantees.
+It likely has a more verbose diagnostic API that describes its state, which should be a view over the results of these commands.
+
+*   `Create(id, description) -> (ReadCapability, WriteCapability)`: binds to `id` the TVC described by `description`.
+
+    The command returns capabilities naming `id`, with frontiers set to initial values chosen by the Storage layer.
+
+    The collection associated with `id` is based on `description`, but it is up to the Storage layer to determine what the source's contents will be.
+    A standarad example is a Kafka topic, whose contents are "added" to the underlying collection as they are observed.
+    The Storage layer only needs to ensure that it can reliably produce the updates between `since` and `upper`, and otherwise the nature of the content is up for discussion.
+    It is an error to re-use a previously used `id`.
+
+*   `SubscribeAt(ReadCapability(id, frontier))`: returns a snapshot of `id` at `frontier`, followed by an ongoing stream of subsequent updates.
```

Comment:
I've been working under the assumption that, from the product side, we'll want the ability to firewall subscribers off for the following sort of reasons:

- A staging dataflowd that talks to production storage to experiment with new view definitions
- Blue/green qualification of new releases (or release candidates) of dataflowd
- Separate business units within a company exporting read-only access to the results of computations with each other (sort of our equivalent of https://docs.snowflake.com/en/user-guide/data-sharing-intro.html)

I'm not sure how to square this with a needing a ReadCapability (which can hold back compaction) to initialize a subscriber. Any thoughts? I can see how a ReadCapability is useful for the production ingestd -> production dataflowd case (so the dataflowd doesn't have to worry too much about falling behind on reading things).

Wild guess: maybe something like a lease with a duration of zero?

Reply:
My take on this is .. foremost, don't read to much into "capability" as implying the policy calls we might make about how they will be used, so much as an internal bookkeeping mechanism. There are multiple principals in a Platform instance, and capabilities seem like the best way to communicate their shared requirements. Handing capabilities out to folks we don't trust is where the leases/timeouts/resource limits show up (e.g. as PostgreSQL does with their WAL).

All that being said: a read capability is a thing that allows you to *start* a subscription, or *start* to maintain a view. It isn't intended to be held by the running instances as its mechanism to perform the read. It is important to hold on to the capability if you'd like to be certain you can restart the subscription at any moment due to a failure, but once started you can still get results from a subscription if you drop your capability / we time it out.

If we want to have low trust, "you get what you get" consumers, giving them a capability with a short time-out (maybe not zero because that's really quite short) and asking them to use `AcquireCapabilities` generally seems to check out. However, that's a pretty crummy experience for many users. Ideally, a human associated with the system would be able to say "our STORAGE compaction doesn't want to lag more than a day behind, for economic reasons" and could hand out leases with day-long leases, knowing that STORAGE can terminate each old lease and kill any stalled subscription, and the only people who feel bad are those that are a day behind. But if set to duration zero, or even "one minute", by us, I could imagine a bunch of frustrated users who can't get their downstream use case up and running.

Reply:
> However, that's a pretty crummy experience for many users.

By which I meant: there is pain associated with this approach, and so we'll want to allow folks to dial around the pain, rather than bake in "high pain". I didn't mean it to be a rhetorical device for "and so, we cannot do this approach".